docs/spec/reactors/consensus/proposer-selection.md

```diff
@@ -2,20 +2,187 @@
+
+Priority Round  | -2 | -1 | 0    | 1   | 2   | 3 | 4 | 5 | Alg step
+--------------- | ---|--- |------|---  |---  |---|---|---|--------
+ |              |    |    |p1,p2 |     |     |   |   |   |Initialized to 0
+ |1             |    |    |      |  p1 |     | p2|   |   |A(i)+=VP(i)
+ |              |    | p2 |      |  p1 |     |   |   |   |A(p2)-= N
+ |2             |    |    |      |     |p1,p2|   |   |   |A(i)+=VP(i)
+ |              | p1 |    |      |     |  p2 |   |   |   |A(p1)-= N
+ |3             |    | p1 |      |     |     |   |   | p2|A(i)+=VP(i)
+ |              |    | p1 |      | p2  |     |   |   |   |A(p2)-= N
+ |4             |    |    |   p1 |     |     |   | p2|   |A(i)+=VP(i)
+ |              |    |    |p1,p2 |     |     |   |   |   |A(p2)-= N
+
+The actual implementation does not create this queue, it only has to store the current accumulated proposer priority. 
+
+### Validator Set Changes
+At each block height the validator set may change. Some of the changes have implications on the proposer selection.
+
+#### Voting Power Change
+Consider again the earlier example and assume that the voting power of p1 is changed to 6:
+
+Validator | p1 | p2 
+----------|--- | ---
+VP        | 4  | 3
+
+Let's also assume that the last round R before this change the proposer priorites were as shown in first row (R-1 end). As it can be seen, the procedure can continue with round 1 at next height without changes as before. 
+
+|Priority Round| -2 | -1 | 0    | 1   | 2   | 3 | 4 | 5 | Comment
+|--------------| ---|--- |------|---  |---  |---|---|---|--------
+| R end        |    | p2 |      |  p1 |     |   |   |   |update VP(p1)
+| 1            |    |    |      |     | p2  |   |   | p1|A(i)+=VP(i)
+|              | p1 |    |      |     | p2  |   |   |   |A(p1)-= N
+
+
+#### Validator Removal
+Consider a new example with set.
+
+Validator | p1 | p2 | p3 |
+--------- |--- |--- |--- |
+VP        | 1  | 2  | 3  |
+
+Let's assume that the last round was R and the proposer priorites were as shown in first row (R end) with their sum being 0. At this point p2 is removed and round 1 of next height runs. At the end of the round (penultimate row) the sum of priorites is -2 (minus the priority of the removed process). 
+ 
+
+|Priority Round |-3 | -2 | -1 | 0  | 1   | 2   | 3 | 4 | 5 | Comment
+|---------------|-- | ---|--- |--- |---  |---  |---|---|---|--------
+| R end         |p3 |    |    |    | p1  | p2  |   |   |   |remove p2
+| 1             |   |    |    | p3 |     | p1  |   |   |   |A(i)+=VP(i)
+|               |   | p1 |    | p3 |     |     |   |   |   |A(p1)-= N
+| *new step*    |   |    | p1 |    | p3  |     |   |   |   |A(i) -= avg
+
+The procedure could continue without modifications. However, it is possible that after a sufficiently large number of modifications in validator set, the priority values would group towards maximum or minimum allowed values.
```

Comment:
In case sum of priorities is 0 at the beginning of every height (round), does this ensure fairness during period of stability?

Reply:
On its own for sure not. But it's a nice invariant (when it holds) and might be useful to simplify a bit a formal proof.

Reply:
Why not? Isn't that just `simpleProposalSelection`, which is fair ?

Reply:
I think the question was if the fact that sum==0 guarantees fairness. What (we claim) makes simpleProposalSelection fair is this: prop = max(A)...that is we do take into account the priority at election. It would be fair even without ensuring sum==0.